compiler/rustc_codegen_rmc/src/codegen/statement.rs

```diff
@@ -485,51 +485,34 @@ impl<'tcx> GotocCtx<'tcx> {
-            _ => Expr::string_constant(
-                "This is a placeholder assertion message; the rust message requires dynamic string formatting, which is not supported by CBMC",
-            ),
-        };
-
-        let loc = self.codegen_span_option(span);
-        let cbb = self.current_fn().current_bb();
-
-        // TODO: is it proper?
-        //
-        // [assert!(expr)] generates code like
-        //     if !expr { panic() }
-        // thus when we compile [panic], we would like to continue with
-        // the code following [assert!(expr)] as well as display the panic
-        // location using the assert's location.
-        let preds = &self.current_fn().mir().predecessors()[cbb];
-        if preds.len() == 1 {
-            let pred: &BasicBlock = preds.first().unwrap();
-            let pred_bbd = &self.current_fn().mir()[*pred];
-            let pterm = pred_bbd.terminator();
-            match pterm.successors().find(|bb| **bb != cbb) {
-                None => self.codegen_assert_false(arg, loc),
-                Some(alt) => {
-                    let loc = self.codegen_span(&pterm.source_info.span);
-                    Stmt::block(
-                        vec![
-                            self.codegen_assert_false(arg, loc.clone()),
-                            Stmt::goto(self.current_fn().find_label(alt), Location::none()),
-                        ],
-                        loc,
-                    )
+        let mut msg = String::from(
+            "This is a placeholder message; RMC doesn't support message formatted at runtime",
+        );
+        if let Some(values) = fargs[0].struct_expr_values() {
+            if !values.is_empty() {
+                if let ExprValue::AddressOf(msg_value) = values[0].value() {
+                    if let ExprValue::Index { array, .. } = msg_value.value() {
+                        if let ExprValue::StringConstant { s } = array.value() {
+                            msg = s.to_string();
+                        }
+                    }
+                } else {
+                    tracing::error!(?values, "Oh no");
                 }
             }
-        } else {
-            self.codegen_assert_false(arg, loc)
-        }
+        };
+
+        self.codegen_fatal_error(&msg, span)
     }
 
-    // By the time we get this, the string constant has been codegenned.
-    // TODO: make this case also use the Stmt::assert_false() constructor
-    pub fn codegen_assert_false(&mut self, err: Expr, loc: Location) -> Stmt {
-        BuiltinFn::CProverAssert.call(vec![Expr::bool_false(), err], loc.clone()).as_stmt(loc)
+    // Generate code for fatal error which should trigger an assertion failure and abort the
+    // execution.
+    pub fn codegen_fatal_error(&self, msg: &str, span: Option<Span>) -> Stmt {
+        let loc = self.codegen_caller_span(&span);
+        Stmt::block(
+            vec![Stmt::assert_false(msg, loc.clone()), Stmt::assume_false(loc.clone())],
```

Comment:
should we `assume(false)` or `abort()`

Reply:
That's a good question. What do you think? I can try it out with abort and see how CBMC behaves.

Reply:
I replaced this by calling `abort()` instead of `assume(false)`, but I still see this message on CBMC output: 
```
aborting path on assume(false) at file <builtin-library-abort> line 6 function abort thread 0
```
So I don't think it will make much difference. I'll use abort() to improve readability.